content/blog/2019-11-19-Bril-phase-selection-and-ordering/index.md

```diff
@@ -0,0 +1,209 @@
+kind of optimization techniques, we need a substantial training set to
+evaluate our ordering methods.
+
+To create this training set, we created a random Bril program generator. Our
+program generator evaluates each decision independently, and doesn't take any
+care to preserve program wide structure. We follow this flow chart at each
+step.
+
+1. First, we randomly chose whether we want to insert a binary operation, a constant, or an id.
+2. Next, given the instruction type, we decide whether it should be a boolean or an integer.
+3. Finally, we choose whether this instruction should write to a new identifier or an already existing identifier. If it's an already existing identifier we uniformly choose one from all the identifiers of the correct type.
+
+One thing that we've ignored so far is control flow. Considering that several
+of our optimizations are control flow optimizations, we clearly need to
+handle them somehow. However, introducing arbitrary control flow is fairly
+tricky. For example, we may construct non-terminating programs, which make
+evaluating them/checking our optimizations for bugs tricky.
+
+Thus, we settled on constructing a Directed Acyclic Graph. Every fixed
+interval, we insert a label. In addition to that label, we have a chance of
+adding a jump/branch to a succeeding label.
+
+We note that there are a lot of hand set probabilities in this generation
+process that have the ability to heavily affect the types of programs
+generated. For example, introducing a lot of control flow makes it more
+difficult for other optimizations to be applied. Thus, we select a mix of
+random constants.
+
+
+## Phase Ordering Analysis
+So, first of all, how much does the phase ordering even matter? To evaluate
+the effectiveness of our optimizations, we will use # of instructions as a
+proxy.
+
+We will use three different test sets for this. One where we have a lot of
+control flow, one with no control flow, and one with intermediate control
+flow. These will be called CFLow, CFMid, and CFHigh.
+
+Then, we apply 200 random permutations of the 7 optimizations mentioned
+above. We record the worst and the best performing permutation, and report
+that here.
+
+
+| Dataset   |      Avg Low      |  Avg High | Avg Ratio |
+|----------|-------------|------|----|
+| CFLow | 18.7 | 26.3 | .72 |
+| CFMid | 34.0   | 39.0| .87|
+| CFHigh | 52.7 | 54.7| .96 |
+
+As we can see, especially on the programs with low amounts of control flow,
+there can be a huge difference between the best and worst permutation.
+
+However, merely the min and max does not provide much information about how
+the distribution of possible permutations performs. To show that, we will
+normalize all permutation performances by the worst performing permutations
+and plot the distribution as a violin plot. We see that there is a tail of
+permutations that performs substantially better than the majority of
+permutations, many of which perform as bad as the worst permutation.
```

Comment:
Cool; that's an interesting finding!

Reply:
Yeah I was pretty surprised by that too - especially since the worst permutation still performs fairly well.

When I looked into a couple of examples, it seemed like there were just 2 classes of optimization opportunities: those generated by the original program and those that were enabled by other optimizations. The worst permutations didn't allow for any interactions between optimizations, they just did the optimizations that were enabled by the original program.

Reply:
Makes sense—so in this set of optimizations, the "hierarchy" of interactions between optimizations collapsed to just a single level.